### DISCUSSION TITLE
How to visualize providers function call paths?

### DISCUSSION THREAD

**Discussion by xnox** [openssl/openssl#24724] [Q&A]
When working on providers code, often the implementation of default & fips providers are co-located with the libcrypto/libssl code itself. It is quite difficult to understand the call traces and which functions call what.

Is there a good way to visualize providers framework and expected call paths?

(either pre-generated cyclic graphs or something one can call to generate)

Specifically something that can visualize / produce call traces for things like RAND SEED initialization as being discussed in https://github.com/openssl/openssl/pull/24057#discussion_r1610707392 as even with detailed pointers to code, I am still failing to grasp the whole model in my head.

**Comment by nhorman** [comment]
The strategic architecture can be found here:
https://www.openssl.org/docs/OpenSSLStrategicArchitecture.html

If you are looking for specific call stack traces, I'm afraid there is no current flow graph available that I am aware of

**Comment by paulidale** [comment]
That te call chain can get convoluted is an understatement.  The FIPS provider is the ideal case for a loaded, rather than internal, provider because it is completely self contained.  Once the FIPS provider is active, the only way to get back to libcrypto is via the call backs which are known.  I don't think we've got any situations where control flow reenters the FIPS provider after a call back.

Nonetheless, it is still really difficult to debug code in the FIPS provider.  The major problems I've experienced are:

1. The question: _is this function from libcrypto or the FIPS provider?_ is often difficult to resolve.  Looking at the raw address will resolve this but it takes some discipline to remember to do this.
2. Looking into structures is the biggest issue I've encountered because both libcrypto and the FIPS provider define more than a few structures with the same name but different fields.
3. It can be quite confusing working out if this is the FIPS provider's internal library context or the libcrypto library context the operation is happening within.

The RNG path is one of the worst with respect to figuring out which function is which and which library context is being used.  The latter causing the need for #21964 for example.
